server/benchmarkmodel/serializers.py

```diff
@@ -114,3 +114,60 @@ def update(self, instance, validated_data):
             instance.approved_at = timezone.now()
         instance.save()
         return instance
+
+
+class ModelPrioritySerializer(serializers.Serializer):
+    priority_rank = serializers.IntegerField(write_only=True)
+    updated_models_order = serializers.JSONField(read_only=True)
+
+    def validate(self, data):
+        if not self.instance:
+            raise serializers.ValidationError("No model association found")
+        if self.instance.approval_status != "APPROVED":
+            raise serializers.ValidationError(
+                "The association should be approved before modifying its priority"
+            )
+        if data["priority_rank"] < -1 or data["priority_rank"] == 0:
+            raise serializers.ValidationError(
+                "The priority value should be a positive integer or -1"
+            )
+        return data
+
+    def update(self, instance, validated_data):
+        priority_rank = validated_data["priority_rank"]
+        bmk_models = BenchmarkModel.objects.filter(
+            benchmark__id=self.instance.benchmark.id
+        )
+        priorities = bmk_models.values_list("priority", flat=True)
+
+        if priority_rank == -1 or priority_rank >= priorities.count():
+            instance.priority = priorities[-1] + 1.0
+        elif priority_rank == 1:
+            instance.priority = priorities[0] - 1.0
+        else:
+            priority1 = priorities[priority_rank - 1]
+            priority2 = priorities[priority_rank - 2]
+            instance.priority = (priority1 + priority2) / 2
+        instance.save()
+        self._rescale_priorities(priorities)
+        return instance
+
+    def _rescale_priorities(self, priorities):
+        """
+        Prevents floating-point underflow by early rescaling priority values
+        """
+
+        differences = [
+            priorities[i + 1] - priorities[i] for i in range(priorities.count() - 1)
+        ]
+        # The number is chosen arbitrarily
+        if min(differences) >= 1e-10:
+            return
+
+        bmk_models = BenchmarkModel.objects.filter(
+            benchmark__id=self.instance.benchmark.id
+        )
+        for i, bmk_model in enumerate(bmk_models):
+            bmk_model.priority = i + 1.0
+
+        BenchmarkModel.objects.bulk_update(bmk_models, ["priority"])
```

Comment:
@johnugeorge Do you suggest having this logic on the client rather than the server?

Reply:
much easier and more REST-friendly to move this logic to the client.